src/Compilers/VisualBasic/Portable/Symbols/Source/CustomModifierUtils.vb

```diff
@@ -2,9 +2,65 @@
     Friend Class CustomModifierUtils
+
+        ''' <summary>
+        ''' sourceMethod has the custom modifiers
+        ''' </summary>
+        Friend Shared Sub CopyMethodCustomModifiers(
+                                sourceMethod As MethodSymbol,
+                                destinationTypeParameters As ImmutableArray(Of TypeSymbol),
+                                <[In], Out> ByRef destinationReturnType As TypeSymbol,
+                                containingAssembly As AssemblySymbol,
+                                <[In], Out> ByRef parameters As ImmutableArray(Of ParameterSymbol))
+
+            Debug.Assert(sourceMethod IsNot Nothing)
+
+            ' For the most part, we will copy custom modifiers by copying types.
+            ' The only time when this fails Is when the type refers to a type parameter
+            ' owned by the overridden method.  We need to replace all such references
+            ' with (equivalent) type parameters owned by this method.  We know that
+            ' we can perform this mapping positionally, because the method signatures
+            ' have already been compared.
+            Dim constructedMethod As MethodSymbol = sourceMethod.ConstructIfGeneric(destinationTypeParameters)
+
+            parameters = CustomModifierUtils.CopyParameterCustomModifiers(constructedMethod.Parameters, parameters)
+
+            Dim destinationReturnTypeWithCustomModifiers = constructedMethod.ReturnType
+
+            ' We do an extra check before copying the return type to handle the case where the overriding
+            ' method (incorrectly) has a different return type than the overridden method.  In such cases,
+            ' we want to retain the original (incorrect) return type to avoid hiding the return type
+            ' given in source.
+            If destinationReturnType.IsSameType(destinationReturnTypeWithCustomModifiers, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) Then
+                destinationReturnType = CopyTypeCustomModifiers(destinationReturnTypeWithCustomModifiers, destinationReturnType, containingAssembly)
+            End If
+
+        End Sub
+
+        ''' <summary>
+        ''' sourceType has the custom modifiers
+        ''' </summary>
+        Friend Shared Function CopyTypeCustomModifiers(sourceType As TypeSymbol, destinationType As TypeSymbol, containingAssembly As AssemblySymbol) As TypeSymbol
+            Dim resultType As TypeSymbol
+
+            Debug.Assert(sourceType.IsSameType(destinationType, TypeCompareKind.AllIgnoreOptionsForVB))
+
+            If destinationType.ContainsTuple() AndAlso Not sourceType.IsSameType(destinationType, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) Then
+                Dim names As ImmutableArray(Of String) = VisualBasicCompilation.TupleNamesEncoder.Encode(destinationType)
+                resultType = TupleTypeDecoder.DecodeTupleTypesIfApplicable(sourceType, containingAssembly, names)
```

Comment:
I looked at implementation of TupleTypeDecoder and, I believe, this is an unnecessary dependency. All necessary information is already available inside array type symbols that are being transformed. Please open an issue for this.  

---
In reply to: [82482088](https://github.com/dotnet/roslyn/pull/14229#discussion_r82482088) [](ancestors = 82482088,82471107)